src/scanner.rs

```diff
@@ -0,0 +1,262 @@
+use crate::token::{Token, TokenType};
+use crate::util::*;
+
+use std::iter::Peekable;
+use unicode_segmentation::{Graphemes, UnicodeSegmentation};
+
+pub struct Scanner<'a> {
+    pub source: Peekable<Graphemes<'a>>,
+    pub tokens: Vec<Token>,
+
+    pub line: usize,
+}
+
+impl<'a> Scanner<'a> {
+    pub fn new(s: &'a str) -> Self {
+        Self {
+            source: UnicodeSegmentation::graphemes(s, true).peekable(),
+            line: 0,
+            tokens: Vec::new(),
+        }
+    }
+
+    fn error(&self, message: &str) {
+        error(self.line, message);
+    }
+
+    fn add_token(&mut self, token_type: TokenType) {
+        self.tokens.push(Token {
+            token_type,
+            line: self.line,
+        })
+    }
+
+    fn peek(&mut self) -> &str {
+        let c = self.source.peek();
+
+        if c.is_none() {
+            EOF
+        } else {
+            *c.unwrap()
+        }
+    }
+
+    fn consume(&mut self) {
+        let c = self.source.next();
+
+        if c.is_none() {
+            error(self.line, "unexpected end of file");
+        }
+    }
+
+    fn consume_and_return(&mut self) -> &str {
+        let c = self.source.next();
+
+        if c.is_none() {
+            error(self.line, "unexpected end of file");
+        }
+
+        c.unwrap()
+    }
+
+    fn consume_if_next(&mut self, condition: fn(&str, &str) -> bool) -> bool {
+        if self.peek() == EOF {
+            false
+        } else {
+            let mut cloned = self.source.clone();
+            let c1 = match cloned.next() {
+                None => EOF,
+                Some(s) => s,
+            };
+
+            let c2 = match cloned.next() {
+                None => EOF,
+                Some(s) => s,
+            };
+
+            if condition(c1, c2) {
+                self.consume();
+            }
+            return condition(c1, c2);
+        }
+    }
+
+    fn consume_and_return_if(&mut self, condition: fn(&str) -> bool) -> Option<&str> {
+        let n = self.peek();
+        if n == EOF || !condition(n) {
+            None
+        } else {
+            Some(self.consume_and_return())
+        }
+    }
+
+    fn consume_while(&mut self, condition: fn(&str) -> bool) {
+        loop {
+            let n = self.peek();
+            if n == EOF || !condition(n) {
+                break;
+            }
+            self.consume()
+        }
+    }
+
+    fn scan_string(&mut self) {
+        let mut out = String::new();
+        let mut lines = 0;
+        loop {
+            if self.peek() == EOF {
+                self.error("unterminated string literal");
+                return;
+            }
+            let n = self.consume_and_return();
+
+            if n == "\n" {
+                lines = lines + 1;
+            }
+            if n == "\"" {
+                break;
+            }
+            out += n;
+        }
+
+        self.add_token(TokenType::String(out))
+    }
+
+    fn scan_number(&mut self) {
```

Comment:
not a very good number parser, but making a good one is really hard

this will fail to roundtrip a lot of floats and will break horribly on out-of-range floats and you'd probably want to 1) keep all precision 2) truncate out-of-range numbers properly